src/emulator/storage/files.ts

```diff
@@ -216,13 +232,18 @@ export class StorageLayer {
     return this._uploads.get(uploadId);
   }
 
-  public cancelUpload(uploadId: string): ResumableUpload | undefined {
-    const upload = this._uploads.get(uploadId);
-    if (!upload) {
-      return undefined;
+  /**
+   * Deletes partially uploaded file from persistence layer and updates its status. Cancelling
+   * an upload is idempotent.
+   * @param upload The upload to be cancelled.
+   * @returns Whether the upload was cancelled (i.e. its initial status was ACTIVE or CANCELLED).
+   */
+  public cancelUpload(upload: ResumableUpload): boolean {
+    if (upload.status === UploadStatus.ACTIVE) {
+      this._persistence.deleteFile(upload.fileLocation);
+      upload.status = UploadStatus.CANCELLED;
     }
-    upload.status = UploadStatus.CANCELLED;
-    this._persistence.deleteFile(upload.fileLocation);
+    return upload.status === UploadStatus.CANCELLED;
```

Comment:
Wouldn't this always return true? If it was ACTIVE, it will be set to `CANCELLED` a few lines above. If it was `CANCELLED`, it stays that way.

Reply:
It can also be `FINISHED`, in which case we want to return a 400.